### PR TITLE
Prevent SigTerms/SigText from running on fields they do not support

### PR DESCRIPTION
**Pull request by polyfractal**
SigTerms cannot run on fields that are not searchable, and SigText cannot run on fields that do not have analyzers.  Both of these situations fail today with an esoteric exception, so this just formalizes
the constraint by throwing an IllegalArgumentException up front.

In practice, the only affected field seems to be the `binary` field, which is neither searchable or has a default analyzer (e.g. even numeric and keyword fields have a default analyzer despite not being tokenized).  Theoretically SigText should probably only run on `text` fields, but it _technically_ works on anything with an analyzer today and changing that behavior is the purview of a different PR :)

Adds supported-type tests, and makes some changes to the test itself to allow testing sigtext (indexing _source).

Also a few tweaks to the test to avoid bad randomization (negative numbers, etc).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTextAggregatorFactory.java [hunk 1]
```diff
@@ -83,6 +83,10 @@ public SignificantTextAggregatorFactory(String name,
         // Note that if the field is unmapped (its field type is null), we don't fail,
         // and just use the given field name as a placeholder.
         this.fieldType = queryShardContext.fieldMapper(fieldName);
+        if (fieldType != null && fieldType.indexAnalyzer() == null) {
+            throw new IllegalArgumentException("Field [" + fieldType.name() + "] has no index analyzer, but SignificantText " +
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTextAggregatorFactory.java]: s/index analyzer/analyzer/? Most folks don't declare a different index and search analyzers so they'll probably be confused about this message.

Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTextAggregatorFactory.java]: Yeah that's fair.  _Technically_ it's the index analyzer that is fetched explicitly, but I could see that causing confusion.  Will change!

**Comment by polyfractal** [comment]
Gonna kick off another CI since it might be a bit stale atm.

@elasticmachine update branch

**Review by nik9000 — APPROVED** [review]
Lgtm
